### PR TITLE
Predictable `QueryBuilder::executeQuery()` and `QueryBuilder::executeStatement()`

### LINKED ISSUE #4461: Introduce fetch* methods in QueryBuilder for SELECT queries
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes
| RFC         | yes
| BC Break    | no

#### Summary

Since DBAL deprecates concept of fetch mode in favor usage specific to every fetch mode Connection methods, there are no easy way to avoid deprecated staff with QueryBuilder usage.

As an alternative to deprecated API DBAL may provide  new methods in QueryBuilder - `fetchAllAssociative()` and other, related to new fetch modes.

Implementation for fetch associative method in QueryBuilder may looks like this:
 
```php
public function fetchAllAssociative(): array
{
    // assert current state contains SELECT query
 
    return $this->connection->fetchAllAssociative($this->getSQL(), $this->params, $this->paramTypes);
}
```

### PR DESCRIPTION
**Pull request by PowerKiKi**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no
| Fixed issues | https://github.com/doctrine/dbal/issues/4461#issuecomment-778875457

#### Summary

This deprecates `QueryBuilder::execute()`, because its return type is
unpredictable and raises issues with static analysis tools such as PHPStan.

Instead you should use either `QueryBuilder::executeQuery()` or
`QueryBuilder::executeStatement()`, depending on whether the queryBuilder is a query (SELECT)
or a statement (INSERT, UPDATE, DELETE).

You might also consider the use of the new shortcut methods, such as:

- `fetchAllAssociative()`
- `fetchAllAssociativeIndexed()`
- `fetchAllKeyValue()`
- `fetchAllNumeric()`
- `fetchAssociative()`
- `fetchFirstColumn()`
- `fetchNumeric()`
- `fetchOne()`

This commit is a direct follow-up to https://github.com/doctrine/dbal/issues/4461
where those shortcut methods where introduced.

### CODE REVIEW

**Comment by beberlei** [comment]
If we do this change, then we could look to add this `Statement` in general maybe even in 2.x, because it would help us solve the issue with `Statement::execute()` returning a boolean at the moment. The API could be in general `Statement::executeStatement` (yeah naming) and `Statement::executeQuery`.

**Comment by morozov** [comment]
> yeah naming

The naming problem is not on our end. The "statement" on the left-hand side is a _prepared_ statement. The "statement" on the right-hand side is a SQL _statement_ (DDL/DML/DCL) as opposed to an SQL _query_. Having it named as [`Statement::executeUpdate()`](https://docs.oracle.com/javase/7/docs/api/java/sql/Statement.html#executeUpdate(java.lang.String)) would be worse IMO.

**Comment by morozov** [comment]
Looks good to me. @beberlei do you want to backport it to `2.x` or it's not necessary for the upgrade path improvement?

**Comment by beberlei** [comment]
@morozov the query builder change is ok for 3.1.x I believe, since the API for QueryBuilder::execute() does not return a boolean, it is already ok to use, would be a new migration step in the future for users. I don't believe there is pressure moving this to 2.13

**Comment by beberlei** [comment]
#4580 would be the one that is more interesting for the 2.x branches

**Review by morozov — COMMENTED** [review]
@PowerKiKi please add a runtime deprecation using `Deprecation::trigger()`.

**Comment by PowerKiKi** [comment]
I added the `Deprecation::trigger()` call as required.

File: src/Query/QueryBuilder.php [hunk 1]
```diff
@@ -335,6 +335,13 @@ public function executeStatement(): int
      */
     public function execute()
     {
+        Deprecation::trigger(
+            'doctrine/dbal',
+            'https://github.com/doctrine/dbal/pull/4578',
+            'QueryBuilder::execute is deprecated, instead use QueryBuilder::executeQuery (for SELECT) or' .
```
Comment (morozov) [inline src/Query/QueryBuilder.php]: This could be moved under the `if` statement and made more specific (without the "or").

Comment (PowerKiKi) [reply src/Query/QueryBuilder.php]: Good idea, I pushed it

Comment (greg0ire) [reply src/Query/QueryBuilder.php]: You should run `vendor/bin/phpcbf`

Comment (PowerKiKi) [reply src/Query/QueryBuilder.php]: done

File: src/Query/QueryBuilder.php [hunk 2]
```diff
@@ -300,19 +300,57 @@ public function fetchFirstColumn(): array
...
      *
      * @throws Exception
      */
     public function execute()
     {
         if ($this->type === self::SELECT) {
+            Deprecation::trigger(
+                'doctrine/dbal',
+                'https://github.com/doctrine/dbal/pull/4578',
+                'QueryBuilder::execute is deprecated, instead use QueryBuilder::executeQuery'
```
Comment (morozov) [inline src/Query/QueryBuilder.php]: Please reference methods using parentheses like it's done in other places. I'd rephrase it to:
```
QueryBuilder::execute() is deprecated, use QueryBuilder::executeQuery() for SQL queries instead.
```

File: src/Query/QueryBuilder.php [hunk 3]
```diff
@@ -300,19 +300,57 @@ public function fetchFirstColumn(): array
...
+            );
+
             return $this->connection->executeQuery($this->getSQL(), $this->params, $this->paramTypes);
         }
 
+        Deprecation::trigger(
+            'doctrine/dbal',
+            'https://github.com/doctrine/dbal/pull/4578',
+            'QueryBuilder::execute is deprecated, instead use QueryBuilder::executeStatement'
```
Comment (morozov) [inline src/Query/QueryBuilder.php]:
```
QueryBuilder::execute() is deprecated, use QueryBuilder:: executeStatement() for SQL statements instead.
```

File: src/Query/QueryBuilder.php [hunk 4]
```diff
@@ -300,19 +300,57 @@ public function fetchFirstColumn(): array
         return $this->connection->fetchFirstColumn($this->getSQL(), $this->params, $this->paramTypes);
     }
 
+    /**
+     * Executes an SQL query (SELECT) and returns a Result.
```
Comment (morozov) [inline src/Query/QueryBuilder.php:304]: This may be misleading since it works also for `DESCRIBE`, `SHOW` and so on. Let's rephrase to:
```
Executes an SQL query that returns a Result.
```

File: src/Query/QueryBuilder.php [hunk 5]
```diff
@@ -300,19 +300,57 @@ public function fetchFirstColumn(): array
...
+     */
+    public function executeQuery(): Result
+    {
+        return $this->connection->executeQuery($this->getSQL(), $this->params, $this->paramTypes);
+    }
+
+    /**
+     * Executes an SQL statement and returns the number of affected rows.
+     *
+     * Should be used for INSERT, UPDATE and DELETE
```
Comment (morozov) [inline src/Query/QueryBuilder.php:316]: It's not limited to DML (`INSERT`, etc.), it could be also DDL and other non-DQL statements. See [JDBC](https://docs.oracle.com/javase/7/docs/api/java/sql/Statement.html) for reference.

**Review by morozov — COMMENTED** [review]
Looks good overall. Please squash at the end.

File: src/Query/QueryBuilder.php (hunk 4 above)
Comment (PowerKiKi) [reply src/Query/QueryBuilder.php:304]: I'd agree with that on the `Connection` class, but AFAIK, there is no way to build anything else than a `SELECT` query with a `QueryBuilder`. So I don't think we should document an impossible use-case.

WDYT ?

see: https://github.com/doctrine/dbal/blob/461b3b70012182777977ad8bef105f1cfaeedcf9/src/Query/QueryBuilder.php#L40-L46

File: src/Query/QueryBuilder.php (hunk 5 above)
Comment (PowerKiKi) [reply src/Query/QueryBuilder.php:316]: Idem, the `QueryBuilder` cannot build anything else than DML. Should we really mention that ?

File: src/Query/QueryBuilder.php (hunk 2 above)
Comment (PowerKiKi) [reply src/Query/QueryBuilder.php]: done

File: src/Query/QueryBuilder.php (hunk 3 above)
Comment (PowerKiKi) [reply src/Query/QueryBuilder.php]: done

File: src/Query/QueryBuilder.php (hunk 4 above)
Comment (morozov) [reply src/Query/QueryBuilder.php:304]: You're right, I didn't take the scope into account. Sorry for the confusion.

File: src/Query/QueryBuilder.php (hunk 5 above)
Comment (morozov) [reply src/Query/QueryBuilder.php:316]: No, you're right.

**Review by morozov — APPROVED** [review]

**Comment by morozov** [comment]
Thanks, @PowerKiKi.
